src/report/bulk.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+/// The parts necessary to take an in-memory
+/// representation of a causal log and produce
+/// a bulk report wire representation.
+///
+/// Note that this is *not* the on-the-wire
+/// representation, but rather an intermediate
+/// helper that can be used to make said data.
+pub struct BulkReportSourceComponents<'log> {
+    /// Where the log data was created/about
+    pub location_id: TracerId,
+    /// The compact log of events and clocks
+    pub log: &'log [CompactLogItem],
+}
+
+impl<'data> BulkReporter for DynamicHistory<'data> {
+    #[inline]
+    fn report_with_extension(
+        &mut self,
+        destination: &mut [u8],
+        extension_metadata: ExtensionBytes<'_>,
+    ) -> Result<usize, ReportError> {
+        if self.chunked_report_state.is_report_in_progress() {
+            return Err(ReportError::ReportLockConflict);
+        }
+        let log = self.compact_log.as_slice();
+        let r = BulkReportSourceComponents {
+            location_id: self.tracer_id,
+            log,
+        }
+        .report_with_extension(destination, extension_metadata);
+        match r {
+            Ok(v) => {
+                self.finished_report_logging();
+                Ok(v)
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+/// Fixed-sized always-initialized header portion of the bulk report format
+#[repr(C, align(1))]
+pub struct WireBulkHeader {
+    /// A magical (constant) value used as a hint about the
+    /// data encoded in this pile of bytes.
+    pub fingerprint: [u8; 4],
+    /// A u32 representing the tracer_id (a.k.a. location id) of the
+    /// ekotrace agent instance producing this report.
+    pub location_id: [u8; 4],
+    /// How many of the payload bytes are populated with log data?
+    pub n_log_bytes: [u8; 4],
+    /// How many of the payload bytes are populated with extension data?
+    pub n_extension_bytes: [u8; 4],
+}
+const_assert_eq!(1, align_of::<WireBulkHeader>());
+const_assert_eq!(16, size_of::<WireBulkHeader>());
+
+/// Attempt to split a bulk report from its on-the-wire representation
+/// into its constituent parts, without unbounded copying or any allocation.
+///
+/// Returns the source location id, the log payload bytes, and the extension payload bytes.
+/// The log payload bytes are expected to be interepreted as little-endian `CompactLogItem`s.
+/// Payload alignment is not addressed.
+pub fn try_bulk_from_wire_bytes(
+    wire_bytes: &[u8],
+) -> Result<(TracerId, &[u8], ExtensionBytes), ParseBulkFromWireError> {
+    if wire_bytes.len() < size_of::<WireBulkHeader>() {
+        return Err(ParseBulkFromWireError::MissingHeader);
+    }
+    assert_eq_align!(u8, WireBulkHeader);
+    debug_assert_eq!(
+        0,
+        wire_bytes.as_ptr() as usize % align_of::<WireBulkHeader>()
+    );
+    let (header_bytes, payload_bytes) = wire_bytes.split_at(size_of::<WireBulkHeader>());
+    let wire_header = unsafe { &*(header_bytes.as_ptr() as *const WireBulkHeader) };
```

Comment:
While I don't see anything specifically wrong here, this pattern is one that is known to be fragile in many dimensions. Perhaps a binary parser using nom could be more appropriate?

Reply:
Trying quite hard to keep the dependencies extremely low here. Is there a particular fragility or set thereof you're worried about needing remediation for?

Reply:
Dimension 1: Binary protocol / file format parsing code is a very common source of security vulnerabilities in networking code and imaging / video libraries (which are effectively network facing). 

Dimension 2: The practice of coercing a byte buffer into a struct is fragile, as it relies on the struct being laid out in memory just-so. You've done the necessary things to make it so (:picard:), but it relies on compiler behavior that is outside of mainstream. 

Dimension 3: The struct definition itself is a source of fragility; when someone else comes along, it may not be obvious why the struct must be laid out just-so, or that alteingr its contents is verboten. 

The only argument to really be made for struct-coercion is that there is absolutely nothing faster. But in this case, you're already (correctly) going through endian conversion code after the coercion. So I don't think that's what you're going after here, and I don't think that it's warranted in this case anyway. 

Whether we use Nom for it or not, we should probably segregate the wire-protocol code into a separate module with its own (crate-internal) API, where we can easily test the snot out of it with very well-defined boundaries. Definitely round-trip proptests from the API side, and preferably fuzz tests from the binary side.

Reply:
Definitely agree on even more testing. There are some round-trip proptests over in the allog_log_report bit, but I'm positive there's room for improvement. 

I'll add more, regardless.

As you may have surmised, the main reasons for going with the struct overlay approach initially here were:
* No dependencies required
* Speed of development
* Compiler assistance re: structs as executable specifications. 
* Not much room for unexpected performance costs to hide

 I don't know what to say about "relies on compiler behavior that is outside of mainstream" -- this is a Rust project, and it seems reasonable to rely on the features of the compiler regarding collections of typed fields.  The `repr`s used are stable, and `static_assertion`s help fill in the gaps around comparability, though I'm sure we could add `assert`s to get runtime double-checking around layout if we're feeling suspicious of `static_assertion` as (this project's only real) 3rd party library.